docs/Tools/Barstools.rst

```diff
@@ -3,3 +3,92 @@ Barstools
 
 Barstools is a collection of useful FIRRTL transformations and compilers to help the build process.
 Included in the tools are a MacroCompiler (used to map Chisel memory constructs to vendor SRAMs), FIRRTL transforms (to separate harness and top-level SoC files), and more.
+
+Mapping technology SRAMs (MacroCompiler)
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you are planning on building a real chip, it is likely that you will plan on using some amount of static random access memory, or SRAM.
+SRAM macros offer superior storage density over flip-flop arrays at the cost of restricting the number of read or write transactions that can happen in a cycle.
+Unlike in Verilog, these types of sequential memory elements are first-class primitives in Chisel and FIRRTL (``SeqMem`` elements).
+This allows Chisel designs to contain abstract instantiations of sequential memory elements without knowing the underlying implementation or process technology.
+
+Modern CAD tools typically cannot synthesize SRAMs from a high-level RTL description.
+This, unfortunately, requires the designer to include the SRAM instantiation in the source RTL, which removes its process portability.
+In Verilog-entry designs, it is possible to create a layer of abstraction that allows a new process technology to implement a specific sequential memory block in a wrapper module.
+However, this method can be fragile and laborious.
+
+The FIRRTL compiler contains a transformation to replace the ``SeqMem`` primitives called ``ReplSeqMem``.
+This simply converts all ``SeqMem`` instances above a size threshold into external module references.
+A list of unique SRAM configurations is output to a ``.conf`` file by FIRRTL, which is used to map technology SRAMs.
+Without this transform, FIRRTL will map all ``SeqMem`` s to flip-flop arrays with equivalent behavior.
+With very high capacity ``SeqMem`` s, this may lead to a design that is difficult to route.
+
+The ``.conf`` file is consumed by a tool called MacroCompiler, which is part of the :ref:`Barstools` scala package.
+MacroCompiler is also passed an ``.mdf`` file that describes the available list of technology SRAMs or the capabilities of the SRAM compiler, if one is provided by the foundry.
+Using a user-customizable cost function, MacroCompiler will select the SRAMs that are the best fit for each dimensionality in the ``.conf`` file.
```

Comment:
What is the default cost function and how does one specify a custom cost function?

Reply:
The default is specified below and I now describe how to override it.